### PR TITLE
Support concatenated gzip (mgzip/pigz) layers in ztoc

### LINKED ISSUE #1834: [Bug] All index in ztoc have start_span and end_span 0 even with offset reaching 250mb+
### Description

I am trying to convert an image using basic soci convert with default settings. But when I try to pull the image using soci snapshotter I get this error - 

```
sudo nerdctl run --platform linux/amd64 --snapshotter soci --net host --rm -it --platform linux/amd64 xxx-container-registry.xxx.com/r1/r2:0.1.0-soci /bin/bash

/bin/bash: error while loading shared libraries: /app/var/data/conda-env/lib/libtinfo.so.6: cannot read file data: Input/output error
```

Soci error log - 
```
Jan 08 18:01:20 lima-default soci-snapshotter-grpc[23927]: {"digest":"sha256:f3f9d548535bfa8739e2b6c6464f1ff5c94bbb3ec28cf6xxxx",
"error":"file.Read: error reading tar header at 51757568, size 512: EOF",
"fetchedPercent":0,"fetchedSize":0,
"level":"error","msg":"statFile error","size":89539606}
```

In the ztoc info I see that its at a offset of `199165440` with span 0 whereas it should have been at span 50 assuming 4mb spans? - 

```
    {
      "filename": "app/var/data/conda-env/lib/libtinfo.so.6",
      "offset": 199165440,
      "size": 0,
      "type": "symlink",
      "start_span": 0,
      "end_span": 0
    },
    {
      "filename": "app/var/data/conda-env/lib/libicuuc.so.75",
      "offset": 199165952,
      "size": 0,
      "type": "symlink",
      "start_span": 0,
      "end_span": 0
    },
```

Here is the end of the ztoc, all are at span 0 while the whole uncompressed layer is 250Mb+

```
    {
      "filename": "app/var/data/conda-env/ssl/misc/CA.pl",
      "offset": 255863296,
      "size": 13398,
      "type": "reg",
      "start_span": 0,
      "end_span": 0
    },
    {
      "filename": "app/var/data/conda-env/ssl/misc/tsget.pl",
      "offset": 255877632,
      "size": 6746,
      "type": "reg",
      "start_span": 0,
      "end_span": 0
    },
    {
      "filename": "app/var/data/conda-env/ssl/misc/tsget",
      "offset": 255885312,
      "size": 0,
      "type": "symlink",
      "start_span": 0,
      "end_span": 0
    },
    {
      "filename": "app/var/data/conda-env/ssl/ct_log_list.cnf.dist",
      "offset": 255885824,
      "size": 412,
      "type": "reg",
      "start_span": 0,
      "end_span": 0
    }
```

I have tried creating spans of 8mb as well and I still see start_span as 0 for all the index in ztoc.

More info for both the ztoc created (the issue is with the first one) - 

```
sudo soci ztoc info sha256:59ecf25ca78e1edc5d17bd6d48f77de1d45073c11b2cb86b684abxxxx | jq '{num_spans, num_files, num_multi_span_files, span_size, size, version, build_tool}'
{
  "num_spans": 1,
  "num_files": 9033,
  "num_multi_span_files": 0,
  "span_size": 4194304,
  "size": 1916800,
  "version": "0.9",
  "build_tool": "AWS SOCI CLI v0.2"
}

sudo soci ztoc info sha256:a0beb7eab9f45aa3fb5220b77953db4c19dcf845bad33219f18yyyy | jq '{num_spans, num_files, num_multi_span_files, span_size, size}'
{
  "num_spans": 58,
  "num_files": 10,
  "num_multi_span_files": 2,
  "span_size": 4194304,
  "size": 1908120
}
```
### Steps to reproduce the bug

_No response_

### Describe the results you expected

- Ztoc index to have proper span start and end and respect the span size of 4mb.
- `file.Read: error reading tar header at xxx` to not occur

### Host information

1. OS: Tried both arm64 and amd64
2. Snapshotter Version: v0.12.1 
3. Containerd Version: v2.2.0

### Any additional context or information about the bug

When pulling the image without SOCI everything works fine.

Wrt the error above, I am able to manually read the tar file after downloading it locally - 

```
gunzip -c /tmp/layer.tar.gz | dd bs=512 skip=$((51757568/512)) count=1 2>/dev/null | xxd | head

00000000: 6170 702f 7661 722f 6461 7461 2f63 6f6e  app/var/data/con
00000010: 6461 2d65 6e76 2f6c 6962 2f6c 6962 7469  da-env/lib/libti
00000020: 6e66 6f2e 736f 2e36 2e35 0000 0000 0000  nfo.so.6.5......
00000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000060: 0000 0000 3030 3030 3737 3500 3030 3131  ....0000775.0011
00000070: 3631 3100 3030 3030 3030 3000 3030 3030  611.0000000.0000
00000080: 3130 3733 3531 3000 3030 3030 3030 3030  1073510.00000000
00000090: 3030 3000 3030 3135 3331 3400 3000 0000  000.0015314.0...
```

_No response_

### PR DESCRIPTION
**Pull request by prafgup**
**Issue #, if available:**
Fixes #1834. The C-level zinfo code only processed the first gzip member, causing all ztoc entries to map to span 0 for mgzip-compressed layers. This adds handling for multiple concatenated gzip members in all three functions: index generation, file-based extraction, and buffer-based extraction.

**Description of changes:**

- Handle concatenated gzip members in `generate_zinfo_from_fp` by detecting additional data after Z_STREAM_END and resetting inflate state for the next member
  - Ref - https://github.com/madler/zlib/blob/master/examples/zran.c#L221-L228
- Handle concatenated gzip members in both extraction functions (`extract_data_from_fp`, `extract_data_from_buffer`) by manually skipping the 8-byte gzip trailer on the first member boundary (raw inflate mode), then switching to gzip mode (windowBits=47) which auto-consumes subsequent trailers
  -  Ref - https://github.com/madler/zlib/blob/master/examples/zran.c#L418-L436

**Testing performed:**
- Added unit test - 
```
lima sudo go test -v -count=1 -run TestDecompressWithPigz ./ztoc/...
=== RUN   TestDecompressWithPigz
=== RUN   TestDecompressWithPigz/span_size_64KB
=== RUN   TestDecompressWithPigz/span_size_256KB
=== RUN   TestDecompressWithPigz/span_size_1MB
--- PASS: TestDecompressWithPigz (0.53s)
    --- PASS: TestDecompressWithPigz/span_size_64KB (0.17s)
    --- PASS: TestDecompressWithPigz/span_size_256KB (0.17s)
    --- PASS: TestDecompressWithPigz/span_size_1MB (0.18s)
PASS
ok      github.com/awslabs/soci-snapshotter/ztoc        0.540s
PASS
ok      github.com/awslabs/soci-snapshotter/ztoc/compression    0.002s [no tests to run]
?       github.com/awslabs/soci-snapshotter/ztoc/compression/fbs/zinfo  [no test files]
?       github.com/awslabs/soci-snapshotter/ztoc/fbs/ztoc       [no test files]
```

```
lima sudo GO_TEST_FLAGS="-run TestSociZtocWithEstargzLayers -count=1 -timeout 10m" make integration


 SOCI_SNAPSHOTTER_PROJECT_ROOT=/Volumes/git/prafulg/soci-snapshotter
 === RUN   TestSociZtocWithEstargzLayers
 === PAUSE TestSociZtocWithEstargzLayers
 === CONT  TestSociZtocWithEstargzLayers
 --- PASS: TestSociZtocWithEstargzLayers (10.75s)
 PASS
 ok         github.com/awslabs/soci-snapshotter/integration 102.684s

 DONE 1 tests in 96.664s
```

- Tried converting/indexing the same image layer as https://github.com/awslabs/soci-snapshotter/issues/1834 that threw 0 start and end span, 

```
{
  "version": "0.9",
  "build_tool": "AWS SOCI CLI v0.2",
  "size": 7262136,
  "span_size": 4194304,
  "num_spans": 118,
  "num_files": 15553,
  "num_multi_span_files": 80,
  "files": [
    {
      "filename": "app/var/data/conda-env/x86_64-conda-linux-gnu",
      "offset": 512,
      "size": 0,
      "type": "dir",
      "start_span": 0,
      "end_span": 0
    },
    {
      "filename": "app/var/data/conda-env/x86_64-conda-linux-gnu/bin",
      "offset": 1024,
      "size": 0,
      "type": "dir",
      "start_span": 0,
      "end_span": 0
    },
....
  {
    "filename": "app/var/data/conda-env/lib/libopenblasp-r0.3.30.so",
    "offset": 97394176,
    "size": 41424320,
    "type": "reg",
    "start_span": 23,
    "end_span": 32
  }
....
    {
      "filename": "app/var/data/conda-env/man/man1/bzegrep.1",
      "offset": 498103296,
      "size": 18,
      "type": "reg",
      "start_span": 117,
      "end_span": 117
    },
    {
      "filename": "app/var/data/conda-env/man/man1/bzcmp.1",
      "offset": 498104320,
      "size": 18,
      "type": "reg",
      "start_span": 117,
      "end_span": 117
    },
    {
      "filename": "app/var/data/conda-env/man/man1/bzmore.1",
      "offset": 498105344,
      "size": 4310,
      "type": "reg",
      "start_span": 117,
      "end_span": 117
    }
  ]
}
```

- Manual test - 

```
sudo soci convert --all-platforms xx/functions/user-function:0.2.0 xx/functions/user-function:0.2.0-soci

sudo nerdctl push --all-platforms xx/functions/user-function:0.2.0

-- clean all containerd and soci cache

lima sudo bash -c 'systemctl stop soci-snapshotter-grpc && cp /Volumes/git/prafulg/soci-snapshotter/out/soci-snapshotter-grpc /usr/local/bin/soci-snapshotter-grpc && cp /Volumes/git/prafulg/soci-snapshotter/out/soci /usr/local/bin/soci && systemctl start
  soci-snapshotter-grpc && sleep 2 && systemctl is-active soci-snapshotter-grpc


lima sudo nerdctl run --net host --rm --platform linux/amd64 --snapshotter soci --entrypoint sh xxxx/functions/user-function:0.2.0-soci -c "/app/var/data/conda-env/bin/python3.12 -c \"import sys; print(sys.version)\""
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ztoc/ztoc_test.go [hunk 1]
```diff
@@ -228,6 +228,134 @@ func TestDecompressWithGzipHeaders(t *testing.T) {
 	}
```
Comment (Shubhranshu153) [inline ztoc/ztoc_test.go:229]: needs a e2e test using pigz

File: ztoc/compression/gzip_zinfo.c [hunk 2]
```diff
@@ -471,8 +519,39 @@ int extract_data_from_buffer(void *d, offset_t datalen,
                 ret = Z_DATA_ERROR;
             if (ret == Z_MEM_ERROR || ret == Z_DATA_ERROR)
                 goto extract_ret;
-            if (ret == Z_STREAM_END)
+            if (ret == Z_STREAM_END) {
```
Comment (Shubhranshu153) [inline ztoc/compression/gzip_zinfo.c:523]: <Reading the spec to understand if anything else to consider here>

File: ztoc/ztoc_test.go (hunk 1 above)
Comment (prafgup) [reply ztoc/ztoc_test.go:229]: I'll try to change `TestDecompressConcatenatedGzip` to use `pigz` (from os.exec) instead of using the current `BuildConcatenatedGzipTar` method for compression.

Comment (Shubhranshu153) [reply ztoc/ztoc_test.go:229]: We can add integration test perhaps to test container with actual pigz tests, in some places we might be using some images from ecr public which we can populate with desired test images as needed.

File: ztoc/compression/gzip_zinfo.c [hunk 3]
```diff
@@ -379,8 +391,43 @@ int extract_data_from_fp(FILE *in, struct gzip_zinfo *index, offset_t offset, vo
...
+                /* Handle concatenated gzip member boundary */
+                if (skip || strm.avail_out > 0) {
+                    if (raw_mode) {
+                        /* Skip the 8-byte gzip trailer (CRC32 + ISIZE).
+                           In raw inflate mode (-15), the trailer is NOT
+                           consumed by inflate. After resetting to gzip mode
+                           (47), subsequent trailers are consumed automatically,
+                           so this is only needed once. */
+                        unsigned drop = 8;
```
Comment (Shubhranshu153) [inline ztoc/compression/gzip_zinfo.c]: nit: Magic number. Probably can use some const to define it here. GZIP_TTRAILER

**Comment by Shubhranshu153** [comment]
Overall looks correct to me. Reading more on gzip to see if i am missing something. Would like some e2e tests here though as we dont know if it works correctly through the both creation and execution lifecycle with a pigz compressor.

File: ztoc/ztoc_test.go (hunk 1 above)
Comment (prafgup) [reply ztoc/ztoc_test.go:229]: I think I can get a minimal integration test working, which created a basic docker image locally with pigz compression and uses that to verify that ztoc get-file work fine as well as fuse mounts are present. 

I think for now that should be enough and in future we can add some custom ecr image in `soci-workshop-examples` repository. This would probably needed to be done by some maintainer with access :) 

I will update the PR soon.

Comment (prafgup) [reply ztoc/ztoc_test.go:229]: @Shubhranshu153  I have update the unit test to use pigz for compression as well as added a minimal integration test which generates a pigz based image manually, indexes it and verifies the content of it.

Also I have manually ran and verified the working of the image which was resulting in 0 span ztoc creation in https://github.com/awslabs/soci-snapshotter/issues/1834

Comment (Shubhranshu153) [reply ztoc/ztoc_test.go:229]: we can create a image with pigz in soci-workshop-example and have that test.
Its ok to have feature extension pending for future pr but ideally should have the required test for a new feature.

**Comment by prafgup** [comment]
Hey @Shubhranshu153  just nudging regarding this :) is there something else I should be supporting with here? 😄

**Comment by sondavidb** [comment]
I'll take a look myself as well — sorry this slipped under my radar

File: integration/util_test.go [hunk 4]
```diff
@@ -1040,3 +1043,125 @@ func withContentStoreConfig(opts ...store.Option) snapshotterConfigOpt {
...
+
+type pigzImageInfo struct {
+	ref        string
+	files      map[string]string
+	layerCount int
+}
+
+// buildPigzImage constructs a minimal OCI image with a single pigz-compressed layer
+// and imports it into containerd via "ctr images import"
+func buildPigzImage(t *testing.T, sh *shell.Shell, imageName string) pigzImageInfo {
```
Comment (sondavidb) [inline integration/util_test.go]: This seems correct but also a little long. I wonder if there's any easy way for this to be done for us? nerdctl lets you create an estargz image through their CLI [here](https://github.com/containerd/nerdctl/blob/main/docs/stargz.md#building-stargz-images-using-nerdctl-build) which should also test this use case. I don't know if you need buildkit for it though, if so then yeah this would be a dealbreaker.

Alternatively, we can always use a public pigz/stargz image. It depends on if we expect particularly egregious edge cases that we need to test or not.

**Review by sondavidb — COMMENTED** [review]
Hey @prafgup, LGTM on the changes side but would like a quick comment addressed

File: integration/util_test.go (hunk 4 above)
Comment (prafgup) [reply integration/util_test.go]: Updated to use stargz :) - https://github.com/awslabs/soci-snapshotter/pull/1871/changes#diff-cadfc1d53e6911f61c74f06c99f36e438d4a128a8989101079a4b9d6f6fc864aR480

Not sure if I I can find a trusted pigz/stargz public image, probably later one can upload something manually to `soci-workshop-example` and use that. But I think for now converting a image to stargz and using that seems fine.

**Review by sondavidb — APPROVED** [review]

**Comment by sondavidb** [comment]
Thanks! Let me get 1 more reviewer.

**Review by ayush-panta — APPROVED** [review]
Looks good, nice work.
